### DISCUSSION TITLE
Welcome to nfJson Discussions!

### DISCUSSION THREAD

**Discussion by nfoxdev** [VFPX/nfJson#18] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by costin77** [comment]
Hi, it is possible to create a json string with special characters in it, like this one for example:
{"$type" : "salariat","info","gggggg"}?
more exactly I am referring to "$type" property.

Thank you,
Costin

**Comment by nfoxdev** [comment]
Hello Costin.. nfJson parse to empty objects, so is not possible; 

A new version that can parse using collection will be available soon;
using _o2json / _json2o: 

![image](https://github.com/user-attachments/assets/9845b0b1-16a5-4c32-a87e-3fe7ed7244cd)
_using [nfThemes](https://github.com/nfoxdev/nfThemes) for syntax coloring & nfViewer  ( not released yet )_

**Comment by costin77** [comment]
Hi, it is possible to create a collection like this one:

"sporuriSalariu":
    [
        {
            "isProcent": true,
            "valoare": 21
        },
        {
            "isProcent": false,
            "valoare": 1000
        }
    ],

if yes, please tell me where to look.
Thank you,
Costin

**Reply by nfoxdev** [reply]
just add the open/close brackets:

text to cjson noshow
{ "sporuriSalariu": [ 
{
"isProcent": true,
"valoare": 21
},
{
"isProcent": false,
"valoare": 1000
}
] }
endtext
oTest = nfJsonread( m.cJson )
? oTest.sporuriSalariu(1).valoare && 21

**Comment by costin77** [comment]
I want to be able to create it programmatically (I know how to read it).

**Comment by nfoxdev** [comment]
the easiest way is using a cursor:

* 1
create cursor  temp (  isProcent L,  valoare I )
insert into temp (  isProcent ,  valoare  ) values ( .t.,21 )
insert into temp (  isProcent ,  valoare  ) values ( .f.,1000)

* 2
oTemp1 = nfCursorToObject() 

* 3  the way we want it:
oTemp2 = create('empty')
addproperty(oTemp2,'sporuriSalariu(1)') 
acopy(oTemp1.rows,oTemp2.sporuriSalariu)

* 4 to json..
? nfjsoncreate(m.oTemp2)

**Comment by costin77** [comment]
Myself I used this code:
IF RECCOUNT('csporuriAcordate')>0
	SELECT csporuriAcordate
	osporuriSalariu = createobject('empty')	
	SCAN
		osp = CREATEOBJECT("empty")
		ADDPROPERTY(osp,"isProcent", IIF(csporuriacordate.procent>0,.t.,.f.))
		ADDPROPERTY(osp,"valoare", IIF(csporuriacordate.procent>0,csporuriacordate.procent, csporuriacordate.sumafixa))
		oTipSpor = CREATEOBJECT("empty")
		ADDPROPERTY(oTipSpor,"__type", IIF(EMPTY(csporuriacordate.codrevisal),'sporAngajator','spor') )
		ADDPROPERTY(oTipSpor,"nume", IIF(EMPTY(csporuriacordate.codrevisal), ALLTRIM(csporuriacordate.denumire),ALLTRIM(csporuriacordate.codrevisal)) )
		ADDPROPERTY(osp,"tip",oTipSpor)
		osporuriSalariu.Add(osp)	
	ends
	ADDPROPERTY(oInfo,"sporuriSalariu",osporuriSalariu)
endi

But at the end it creates something like this (notice that i have sporurisalariu_kl_collection instead of sporuriSalariu + an extra 'collectionsitems'):
          "sporurisalariu_kl_collection":{ 
              "collectionitems": [
                  {
                      "isprocent":true,
                      "tip":{
                          "nume":"Alt spor",
                          "$type":"sporAngajator"
                      },
                      "valoare":10.000
                  },
                  {
                      "isprocent":true,
                      "tip":{
                          "nume":"Spor de vechime in munca",
                          "$type":"spor"
                      },
                      "valoare":5.000
                  }
              ]
          },

**Reply by nfoxdev** [reply]
That should not happen if you did not add a collection - 
Share a working code to reproduce the output .

**Comment by costin77** [comment]
Sorry, I am working on my code, instead of:
osporuriSalariu = createobject('empty')
initially I had
osporuriSalariu = createobject('collection')
which generated the output from above.
so I am getting;
"sporurisalariu_kl_collection":{
"collectionitems": [
{"isprocent":true,"tip":{"nume":"Alt spor","$type":"sporAngajator"},"valoare":10.000},
{"isprocent":true,"tip":{"nume":"Spor de vechime in munca","$type":"spor"},"valoare":5.000}
]
},

But what I really need is:
"sporurisalariu": [
{"isprocent":true,"tip":{"nume":"Alt spor","$type":"sporAngajator"},"valoare":10.000},
{"isprocent":true,"tip":{"nume":"Spor de vechime in munca","$type":"spor"},"valoare":5.000}
],

**Reply by nfoxdev** [reply]
Yes, I noticed after reviewing.

In any case, nfJson marks and serialize collections as collectionname_kv_collection ( key/value collection ) and collectionname_kl_collection ( keyless collection )

The output you want should be done using an array, not a collection.

Check the readme; 
Another approach, using  [underscore function](https://github.com/nfoxdev/underscore/tree/master) eases your object creation like this:
```
 * download from https://raw.githubusercontent.com/nfoxdev/underscore/refs/heads/master/_.PRG 
  

  oTest = Createobject('empty')

  With _(m.oTest)

    .osporuriSalariu = .newList()

    Scan

      With .newItemFor('osporuriSalariu')

        .isProcent = csporuriacordate.procent>0
        .valoare   = Evl(csporuriacordate.procent,csporuriacordate.sumafixa)

        With _(.oTipSpor)
          .__type = Iif(Empty(csporuriacordate.codrevisal),'sporAngajator','spor')
          .nume   = Iif(Empty(csporuriacordate.codrevisal), Alltrim(csporuriacordate.denumire),Alltrim(csporuriacordate.codrevisal))
        Endwith

      Endwith

    Endscan

  Endwith

```
